### PR TITLE
Changes to make CPU Manager a Hint Provider for Topology Manager

### PR DESCRIPTION
**Pull request by nolancon**
**What this PR does / why we need it**:
Make CPU Manager a Hint Provider for Topology Manager in order to help make Topology aware resource alignment decisions. 
Based on merged design proposal: kubernetes/community#1680
Issue for tracking PRs: #72828

**What type of PR is this?**

> /kind feature

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by nolancon** [comment]
/sig-node
@lmdaly
/assign @ConnorDoyle  @balajismaniam

**Comment by spiffxp** [comment]
/hold
ref: https://github.com/kubernetes/enhancements/issues/693#issuecomment-466728227

/ok-to-test

File: pkg/kubelet/cm/cpumanager/policy_static.go
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/policy_static.go]: This seems like a strange indirection here. Why not just change `allocateCPUs()` to take a socketMask?

File: pkg/kubelet/cm/cpumanager/policy_static.go
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/policy_static.go]: When is this ever not true?

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/topology_hints.go]: This doesn't seem to have been updated according to my suggestions in: https://docs.google.com/document/d/1UbD1igeBA1sRuqwkQgJ5sJ9sf1rao_LLORWLhwSdGUo

Notes copied from the doc:
```
Checkpointing TopologyHints
As a side note, I had to add some extra logic to the GetTopologyHints() implementation
for the cpumanager to get it to work across kubelet restarts. The main problem being that
the topology information was lost across the restart (since it’s not being checkpointed) and
new hints were being generated for already-running containers. We could either checkpoint
the hint information, or make sure we regenerate valid hints as I do below...
```

**Review by klueska — CHANGES_REQUESTED** [review]

File: pkg/kubelet/cm/cpumanager/policy_static.go
Comment (nolancon) [reply pkg/kubelet/cm/cpumanager/policy_static.go]: Done

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (nolancon) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: Done - may need editing as implemented from notes?

File: pkg/kubelet/cm/cpumanager/policy_static.go
Comment (nolancon) [reply pkg/kubelet/cm/cpumanager/policy_static.go]: sockets would be nil if TopologyHints are nil for the container. Should never really happen, but gives extra security.

File: pkg/kubelet/cm/cpumanager/policy_static.go
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/policy_static.go]: It doesn't seem like this was changed.

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: It says this comment is outdated, but I don't see the code change in the file.

File: pkg/kubelet/cm/cpumanager/policy_static.go
Comment (nolancon) [reply pkg/kubelet/cm/cpumanager/policy_static.go]: Overwrote it again by mistake, should be ok now

File: pkg/kubelet/cm/cpumanager/policy_static.go
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/policy_static.go]: I think the field in the hint is called `SocketAffinity`, not `SocketMask`.

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/topology_hints.go]: We should have a `continue` in this error case or an explicit return (since we only process the "cpu" resource in this function).

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/topology_hints.go]: This feels like it should be more generally usable by the CPUManager as a whole. We should consider a cleanup PR in the future that consolidates this logic into the mainline CPUManager instead of duplicating it here.

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/topology_hints.go]: We don't need to address this now because the current algorithm seems to do the right thing, but we should consider cleaning this logic up a bit.  It seems like the general flow should be:

1) Get a count of how many guaranteed CPUs have been requested: `requested int`
2) Get a count of how many CPUs are available on each socket: `available []int`
3) Build hints from all possible socket combinations where the request can be granted: `hints []TopologyHint`
4) Set hints with the lowest number of sockets set in their socket mask to `Preferred: true`
5) Set all other hints to `Preferred: false`

If we could distill the code in this function down to just 5 function calls that do these 5 things with these comments above them, that would help with understanding the flow quite a bit.

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: This has been resolved in the newest version.

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (cbf123) [inline pkg/kubelet/cm/cpumanager/topology_hints.go]: In the logs I'm getting this:

2019-06-21T21:51:45.041 controller-1 kubelet[2159067]: info I0621 21:51:45.041010 2159067 topology_hints.go:106] [cpumanager] Topology Affinities for pod: [{1000000000000000000000000000000000000000000000000000000000000000 true} {0100000000000000000000000000000000000000000000000000000000000000 true} {1100000000000000000000000000000000000000000000000000000000000000 true}]

First, I think those bitmasks should be the other way around with NUMA node 1 being the rightmost bit instead of leftmost.  Second, I think we should not print the high-order zero bits to minimize what gets printed, since the common cases will likely be 1/2/4 NUMA nodes.

Comment (klueska) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: Reversing the bit order is addressed in https://github.com/kubernetes/kubernetes/pull/78587.

Regarding stripping the high order bits. I would probably not strip them completely. Instead, I would strip them down to lengths of 2, 4, 8, and then multiples of 8 up to 64 depending on which bits are set.

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (moshe010) [inline pkg/kubelet/cm/cpumanager/topology_hints.go]: So I am not sure why cpu manager show give an hint which is preferred = false.  IMHO it should have preferred = true to all of it hints otherwise the following example won't work:
I have 4 VFs 2 VF from numa 0 and 2 VF from numa 1 device manager will return (1,1 preferred = true)
and I requested 2 cpus (I have 2 cpus free on each numa node) cpu manage hints should be [(1,0 preferred = true),(0,1 preferred = true), (1,1 preferred = true). because all 3 option are good. It will be up to the topology manager to find the preferred one of all.

Comment (nolancon) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: If hint providers return as follows:
cpu manager: [(1,0 preferred = true),(0,1 preferred = true), (1,1 preferred = false)]
device manager: [(1,1 preferred = true)]
topology manager will calculate [1,1 preferred = true]

As long as cpu manager returns a better hint eg [1,0] or [0,1] than cross-socket affinity [1,1], then the preferred status of the cross-socket affinity will not influence topology manager's decision. So it is set to false by default with no consequence (as above).

If the cross-socket affinity is indeed the best hint returned by cpu manager, then it will influence the decision of topology manager and should therefore be set to true.
[1,1 preferred = true].
However, in the outlying case that neither numa node could ever satisfy a request individually, but could satisfy a request collectively, the hints returned will be [1,1 preferred = false]. In this case the pod will fail in "strict" mode and pass in "preferred" mode.

Comment (moshe010) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: @nolancon, thank for the reply. 
The is the problem I see in the current logic. 
let say I have 4 vf, 2 VF on NUMA0 and 2 VF on NUMA1 and I have 2 cpus free on NUMA0 and 2 cpus free NUMA1 and we are using strict mode. 
Scenario A:
if a user request in the pod spec 4 VF and 2 cpus
the device plugin manager will return [(1,1 preferred = true)]
the cpu manger will return [(0,1 preferred=true), (1,0 preferred=true) (1,1 preferred = false)]
and we will get Topology Affinity Error.
Scenario B:
if a user request in the pod spec 4 VF and 4 cpus
the device plugin manager will return [(1,1 preferred = true)]
the cpu manager will return [ (1,1 preferred = true)]
A pod will be successfully created.
This seem to me as an  inconsistent behavior.  I would expect both Scenarios to create pod successfully.

Comment (moshe010) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: sorry I was mistaken in  Scenario A we will have a final topology hint (0,1  preferred=true) because on the bitwise and operation. I guess the question is what strict policy means? and if that what the user expected.   Do not  you think that the best hint in that case is (1,1)? so if you have VFs on 2 NUMA don't you want to have cpus on both NUMA has well?

Comment (lmdaly) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: The 1,1 hint indicates that the resources can be taken from either node and delegates the allocation to the hint provider, So in the event there is more device requests or additional hint providers sending on hints we could see resources being allocated from either node - lets say we had another device plugin that could allocate a device from either node and decided to give from numa node 1 and lets say there was a huge page hint provider then it decided to give the resources from numa node 0. 

Whereas in the event that the hint was 0,1 then all hint providers would allocate to the same node where possible and those who can't spread across sockets - these leads to as many resources as possible coming from the one numa node.

Though I understand your case above, maybe we can think more on how to satisfy this case. 

And finally, the strict policy fails in the event that a preferred hint cannot be chosen for any particular resource - lets say in scenario A CPU Manager only returned (1,1 preferred = false) this would fail in strict policy.

Comment (moshe010) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: So maybe the name of the policy is confusing.
Trying to allocate many resources as possible for one numa seem like best effort policy.
Where strict can be one of the following: 
1. allocate all resources one numa 
2. allocate all resources if they have the same socket mask hint
what do you think?

Comment (klueska) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: I agree that there is some confusion in the naming.

At present, there is really only one true "policy" that the `TopologyManager` implements, and that, I agree, this is a best-effort policy. 

The `preferred` vs. `strict`  distinction is actually more of a "rule" for deciding if we admit a pod or not when this best-effort policy is not able to find alignment that includes at least one NUMA node.

Maybe a better name for the policies would be "best-effort-preferred" and "best-effort-strict".

That being said, I am having a hard time picturing how to cleanly support the type of policy that you describe. At present, the SocketAffinity mask in a TopologyHint is designed to be interpreted as an actual _mask_ (i.e. resources can be pulled with affinity to _any_ of the sockets in the mask, not that they *must* be pulled from all of them).

If we attempted to implement a policy similar to what you describe, then each HintProvider would have to somehow know to treat the SocketAffinity mask in the TopologyHint differently (i.e. as a *must* pull resources with affinity to every socket in the mask instead of a pull from *any*).

Intertwining the policy set in the TopologyManager with how to interpret the bits in the SocketAffinity mask by a hint provider seems like a bad idea.

Moreover, I'm not sure how to handle the case where you have (for example) 2 devices that must be allocated cross-NUMA, but you have only requested a single CPU. In your scheme this would block pod admission, I guess, because a single CPU can never be allocated on both NUMA devices at once?

Maybe if we enumerate the exact semantics you would like to see supported, and how you envision the CPUManager and other hint providers doing their allocations based on these semantics, we can come up with a solution that is compatible with the current design.

Comment (moshe010) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: Ok so just to setup the terminology I will call best-effort the algorithm and preferred/strict the rule.

IMO, if the current design is to support just best effort algorithm in this phase then we don't need the preferred/ strict "rule". best effort should always succeed.   

I think that it will be useful that the topology manager  algorithm can be configurable and this can be a later enhancement to the current PR. 

Thought?

Comment (klueska) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: I think I tend to agree.

I've summarized my thoughts around this as one of the points in my latest feedback document: https://docs.google.com/document/d/1g5Aqa0BncQGRedSJH0TJQWq3mw3VxpJ_ufO1qokJ1LE

I also implemented a small python program to compare and contrast the different "merging strategies" we've been discussing for the strict policy. Check it out here:
https://gist.github.com/klueska/02443e7ee37c6d225771ba6fc9c385fc

**Comment by sjenning** [comment]
this currently does not compile with `cm.topologyManager undefined (type *containerManagerImpl has no field or method topologyManager)`

**Comment by moshe010** [comment]
> this currently does not compile with `cm.topologyManager undefined (type *containerManagerImpl has no field or method topologyManager)`
yes it seem that there is a dependency to  https://github.com/kubernetes/kubernetes/pull/74357

**Comment by lmdaly** [comment]
> > this currently does not compile with `cm.topologyManager undefined (type *containerManagerImpl has no field or method topologyManager)`
> > yes it seem that there is a dependency to  #74357

Yep, once #74357 is merged I can rebase this PR and it should pass the tests/compile

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/topology_hints.go]: I get a `govet` error on this line with:
```
Errors from golint:
pkg/kubelet/cm/cpumanager/topology_hints.go:104:22: should drop = 0 from declaration of var totalCPUs; it is the zero value
```

Comment (lmdaly) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: Updated, thanks!

File: pkg/kubelet/cm/cpumanager/policy_static.go
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/policy_static.go]: This logic is currently buggy. It will  only pull CPUs from aligned sockets, regardless of the `TopologyManager` in use. Per offiline discussions, we will go ahead with merging this PR with this bug in place.

I will create an issue explaining the bug in more detail, with a follow-on PR (including some more cleanup) before the release.

Comment (klueska) [reply pkg/kubelet/cm/cpumanager/policy_static.go]: Here is a patch with unit tests:
https://github.com/klueska/kubernetes/commit/c86cb758a87c770717d6938a32767365f00365af

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: I think most of your comments would be addressed by a restructuring according to a previous comment of mine:
https://github.com/kubernetes/kubernetes/pull/73920#discussion_r289394840

The logic was partially reorganized to match this, but not fully.

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (klueska) [inline pkg/kubelet/cm/cpumanager/topology_hints.go]: This could have a state explosion.

Imagine, for example, a 100 CPU system, iterating all possible combinations of requesting 50 CPUs.
That's 1.0089134e+29 combinations we would have to walk through.

We can probably do something smarter here...

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: I have pushed a set of changes that cleanup this section of the code and remove this function.

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: Fixed in latest revision.

File: pkg/kubelet/cm/cpumanager/policy_static.go
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/policy_static.go]: Added to this patch set now in the newest iteration.

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: I have pushed a set of changes that cleanup this section of the code and remove this function.

File: pkg/kubelet/cm/cpumanager/policy_static.go
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/policy_static.go]: I have pushed a set of changes that cleanup this section of the code and remove this function.

File: pkg/kubelet/cm/cpumanager/policy_static.go
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/policy_static.go]: Fixed in latest revision of this PR.

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/topology_hints.go]: Can we make an issue to follow up with this?

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/cpumanager/topology_hints.go]: **NOTE:** It looks like this algorithm is `O(n!)` in number of sockets. 8 sockets already yields ~40k combinations to check. Nowadays IIUC >8 socket systems are pretty exotic, but maybe some kind of guard would help prevent this from crashing/stalling the kubelet.

Comment (ConnorDoyle) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: Adding a couple tests with high socket counts (4 and 8) would also help.

**Review by ConnorDoyle — COMMENTED** [review]
This is looking much better with the recent changes, left just a couple more comments.

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: Yes. The builtin assumption is that the number of sockets is relatively low, so this doesn't explode. I am adding some tests now for 4 and 8 socket system.

In theory, (independent of the CPUManager) we should be generating one hint for every possible socket combination that can satisfy the incoming request (otherwise we miss aligning with an exotic combination from another HintProvider). As such, I don't see a way around enumerating all socket combinations in the worst case (i.e. O(n!)).

I'm all for putting a guard in to ensure the socket count is low enough not to stall the kubelet (I think 8 is a reasonable number to choose too). Where would you recommend putting this guard in? Ideally we would have it fail the kubelet on startup (as opposed to just failing once a pod is submitted), saying we don't support the use of the TopologyManager on systems with > 8 Sockets.

Comment (ConnorDoyle) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: That seems reasonable to me too, but may want to get a Kubelet maintainer to chime in on their preference if they are available to comment.

/cc @derekwaynecarr

Comment (klueska) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: Also, it's not quite O(n!). Its actually:

`sum(n choose k) for k=1...numSockets`

which is much less than n! for large n.

Compare:
```
2 sockets: (n! = 2), (sum(n choose k) = 3)
4 sockets: (n! = 24), (sum(n choose k) = 15)
8 sockets: (n! = 40320),  (sum(n choose k) = 255)
16 sockets: (n! = 20922789888000), (sum(n choose k) = 65535)
```

Comment (klueska) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: Tests added.

Comment (ConnorDoyle) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: Thanks for clarifying the complexity Kevin. Let's follow up with the discussion of guarding topology manager startup on supported number of sockets in an issue and/or upcoming sig-node meeting.

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (bg-chun) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: `future CPUManager polices` does it mean a kind of thread sibling policies?

Comment (ConnorDoyle) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: It’s not talking about any particular new policy. The main point here is that this function implementation breaks the policy abstraction we have set up. It would be preferable to add a new method to the Policy interface and then delegate to that from here instead.

File: pkg/kubelet/cm/cpumanager/topology_hints.go
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: Added to the sig-node meeting for 13-Aug-2019

Comment (ConnorDoyle) [reply pkg/kubelet/cm/cpumanager/topology_hints.go]: Following up, plan discussed for alpha in this release cycle is to fail-fast for large number of socket in the topology manager initialization routine. That will be done in a subsequent PR.

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask_test.go
Comment (ConnorDoyle) [inline pkg/kubelet/cm/topologymanager/socketmask/socketmask_test.go]: nit: interval is [1, 2^n] IIUC.
But, 👍 for this simplification

**Review by ConnorDoyle — APPROVED** [review]
/approve
/lgtm

File: pkg/kubelet/cm/topologymanager/socketmask/socketmask_test.go
Comment (klueska) [reply pkg/kubelet/cm/topologymanager/socketmask/socketmask_test.go]: I'm pretty sure it's the range [1, 2^n-1]:
For example, with 2 sockets we get: {01, 10, 11}
Which has 3 elements (i.e. 2^n -1 elements) representing the numbers 1, 2, and 3 (i.e. [1, 2^n-1])

**Pull request merged**
